cdc/owner/changefeed.go

```diff
@@ -0,0 +1,438 @@
+			c.state.PatchTaskPosition(captureID, func(position *model.TaskPosition) (*model.TaskPosition, bool, error) {
+				return nil, position != nil, nil
+			})
+			ok = false
+		}
+	}
+	for captureID := range c.state.Workloads {
+		if _, exist := captures[captureID]; !exist {
+			c.state.PatchTaskWorkload(captureID, func(workload model.TaskWorkload) (model.TaskWorkload, bool, error) {
+				return nil, workload != nil, nil
+			})
+			ok = false
+		}
+	}
+	return
+}
+
+func (c *changefeed) handleBarrier(ctx cdcContext.Context) (uint64, error) {
+	barrierTp, barrierTs := c.barriers.Min()
+	blocked := (barrierTs == c.state.Status.CheckpointTs) && (barrierTs == c.state.Status.ResolvedTs)
+	if blocked && c.state.Info.SyncPointEnabled {
+		if err := c.sink.SinkSyncpoint(ctx, barrierTs); err != nil {
+			return 0, errors.Trace(err)
+		}
+	}
+	switch barrierTp {
+	case ddlJobBarrier:
+		ddlResolvedTs, ddlJob := c.ddlPuller.FrontDDL()
+		if ddlJob == nil || ddlResolvedTs != barrierTs {
+			c.barriers.Update(ddlJobBarrier, ddlResolvedTs)
+			return barrierTs, nil
+		}
+		if !blocked {
+			return barrierTs, nil
+		}
+		done, err := c.asyncExecDDL(ctx, ddlJob)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		if !done {
+			return barrierTs, nil
+		}
+		c.ddlPuller.PopFrontDDL()
+		newDDLResolvedTs, _ := c.ddlPuller.FrontDDL()
+		c.barriers.Update(ddlJobBarrier, newDDLResolvedTs)
+
+	case syncPointBarrier:
+		if !c.state.Info.SyncPointEnabled {
+			c.barriers.Remove(syncPointBarrier)
+			return barrierTs, nil
+		}
+		if !blocked {
+			return barrierTs, nil
+		}
+		nextSyncPointTs := oracle.GoTimeToTS(oracle.GetTimeFromTS(barrierTs).Add(c.state.Info.SyncPointInterval))
+		c.barriers.Update(syncPointBarrier, nextSyncPointTs)
+
+	case finishBarrier:
+		if !blocked {
+			return barrierTs, nil
+		}
+		c.feedStateManager.MarkFinished()
+	default:
+		log.Panic("Unknown barrier type", zap.Int("barrier type", int(barrierTp)))
+	}
+	return barrierTs, nil
+}
+
+func (c *changefeed) asyncExecDDL(ctx cdcContext.Context, job *timodel.Job) (done bool, err error) {
+	if job.BinlogInfo == nil {
+		log.Warn("ignore the invalid DDL job", zap.Reflect("job", job))
+		return true, nil
+	}
+	cyclicConfig := c.state.Info.Config.Cyclic
+	if cyclicConfig.IsEnabled() && !cyclicConfig.SyncDDL {
+		return true, nil
+	}
+	if c.ddlEventCache == nil || c.ddlEventCache.CommitTs != job.BinlogInfo.FinishedTS {
+		ddlEvent, err := c.schema.BuildDDLEvent(job)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		err = c.schema.HandleDDL(job)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		ddlEvent.Query = binloginfo.AddSpecialComment(ddlEvent.Query)
+		c.ddlEventCache = ddlEvent
+	}
+	if job.BinlogInfo.TableInfo != nil && c.schema.IsIneligibleTableID(job.BinlogInfo.TableInfo.ID) {
+		log.Warn("ignore the DDL job of ineligible table", zap.Reflect("job", job))
+		return true, nil
+	}
+	done, err = c.sink.EmitDDLEvent(ctx, c.ddlEventCache)
```

Comment:
It seems the `*asyncSink` receives a DDL, sends it to its `ddlCh`, and a background routine reads DDL from the `ddlCh` and executes it, so the `done` is always false here, is there any scenario that `done` can be true

Reply:
I will add some comments to async sink